marshmallow_recipe/bake.py

```diff
@@ -132,17 +133,42 @@ def get_field_for(
     if t is Any:
         return raw_field(**metadata)
 
-    if underlying_type_from_optional := _try_get_underlying_type_from_optional(t):
+    underlying_union_types = _try_get_underlying_types_from_union(t)
+    # Optional is a union with None
+    if underlying_union_types is not None and any(t is types.NoneType for t in underlying_union_types):
         required = False
         allow_none = True
-        t = underlying_type_from_optional
     elif metadata.get("default", dataclasses.MISSING) is not dataclasses.MISSING:
         required = False
         allow_none = False
     else:
         required = True
         allow_none = False
 
+    if underlying_union_types is not None:
+        effective_underlying_union_types = [t for t in underlying_union_types if t is not types.NoneType]
+        if not effective_underlying_union_types:
+            raise ValueError("Union must contain at least one type other than NoneType")
+        if len(effective_underlying_union_types) == 1:
+            t = effective_underlying_union_types[0]
+        else:
+            underlying_union_fields = {}
+            for underlying_type in effective_underlying_union_types:
+                effective_underlying_type = get_origin(underlying_type) or underlying_type
+                underlying_union_fields[effective_underlying_type] = get_field_for(
```

Comment:
as I understand `get_origin` was added for usage in `isinstance` but than `list[bool] | list[str]` become unsupported case because `list[str]` will override `list[bool]` with same `list` key in underlying_union_fields and there are no notifications for the user it can be not obvious and unexpected behaviour